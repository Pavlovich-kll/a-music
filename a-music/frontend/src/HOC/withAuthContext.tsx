import React, { createContext, FC, useState } from 'react';

export interface IFormValues {
  username: string;
  firstName: string;
  lastName: string;
  email: string;
  password: string;
  repeatPassword: string;
  phone: string;
  remember: boolean;
}

const INITIAL_VALUES = {
  phone: '',
  username: '',
  firstName: '',
  lastName: '',
  email: '',
  password: '',
  repeatPassword: '',
  remember: false,
};

export const AuthContext = createContext<{ values: IFormValues; setValues: any }>(null);
AuthContext.displayName = 'AuthContext';

export const withAuthContext = (Component: FC) => {
  function WithAuthContext(props: any) {
    const [values, setValues] = useState<IFormValues>(INITIAL_VALUES);

    return (
      <AuthContext.Provider value={{ values, setValues }}>
        <Component {...props} />
      </AuthContext.Provider>
    );
  }
  return WithAuthContext;
};
