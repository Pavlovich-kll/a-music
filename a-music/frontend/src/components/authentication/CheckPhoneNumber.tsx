import React, { useContext } from 'react';
import { Field, Formik, FormikProps } from 'formik';
import { AuthContext } from '../../HOC/withAuthContext';
import { WrappedForm } from './form-styles';
import { Button } from '@material-ui/core';
import { PhoneSchema } from '../../validation-schemas';
import PhoneInput from 'react-phone-input-2';
import 'react-phone-input-2/lib/material.css';
import { toast } from 'react-toastify';
import { IModalChild } from '../../ui/modal';

interface IFormikValues {
  phone: string;
}

interface ICheckPhoneNumber extends IModalChild {
  checkPhoneToProceed(phone: string): void;
}

const CheckPhoneNumber = ({ nextStep, checkPhoneToProceed }: ICheckPhoneNumber) => {
  const { values, setValues } = useContext(AuthContext);

  const handleSubmit = async (formValues: IFormikValues) => {
    try {
      await checkPhoneToProceed(formValues.phone);
      nextStep(true);
    } catch (e) {
      toast.error(e.message);
    } finally {
      setValues({ ...values, ...formValues });
    }
  };

  return (
    <Formik initialValues={values} onSubmit={handleSubmit} validationSchema={PhoneSchema}>
      {({ errors }: FormikProps<IFormikValues>) => (
        <WrappedForm>
          <Field name="phone">
            {({ field, form }: { field: any; form: FormikProps<IFormikValues> }) => {
              const handleChange = (value: string) => form.setFieldValue('phone', `+${value}`);
              return (
                <PhoneInput
                  inputClass={errors.phone ? 'phone-input phone-error' : 'phone-input'}
                  onChange={handleChange}
                  country={'ua'}
                  value={field.value}
                  placeholder="Enter your phone"
                />
              );
            }}
          </Field>
          {errors.phone && <div className="phone-error-text">{errors.phone}</div>}
          <Button fullWidth type="submit" className="phone-check-submit" data-testid="submit-btn">
            Continue
          </Button>
        </WrappedForm>
      )}
    </Formik>
  );
};

export default CheckPhoneNumber;
