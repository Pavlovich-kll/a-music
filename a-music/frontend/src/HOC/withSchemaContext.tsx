import React, { createContext, FC, useReducer, useState } from 'react';
import CheckPhoneNumber from '../components/authentication/CheckPhoneNumber';
import PhoneVerificationSignUp from '../components/authentication/CheckPhoneBySMS';
import Registration from '../components/authentication/Registration';
import Congratulations from '../components/authentication/Congratulations';
import Authenticate from '../components/authentication/Authenticate';
import ChangePassword from '../components/authentication/ChangePassword';
import withCheckNumberRegistration from '../components/authentication/withCheckNumberRegistration';
import withCheckNumberRecovery from '../components/authentication/withCheckNumberRecovery';

export interface IAuthSchemeContext {
  title: string;
  component: string;
  counter: number;
  schemaName: string;
}

interface IAction {
  type: string;
  payload?: boolean;
}

export const INITIAL_STATE = {
  title: '',
  component: '',
  counter: 0,
  schemaName: '',
};

export const AuthSchemeContext = createContext(null);
AuthSchemeContext.displayName = 'AuthSchemeContext';

const PHONE_REG_CHECK_STEP = { title: 'Sign Up', component: withCheckNumberRegistration(CheckPhoneNumber) };
const PHONE_RECOVERY_CHECK_STEP = { title: 'Recover Password', component: withCheckNumberRecovery(CheckPhoneNumber) };
const PHONE_VERIFICATION_STEP = {
  title: 'Phone Verification',
  component: withCheckNumberRecovery(PhoneVerificationSignUp),
};
const REGISTRATION_STEP = { title: 'Registration', component: Registration };
const CONGRATULATIONS_STEP = { title: 'Congratulations', component: Congratulations };
const AUTHENTICATE_STEP = { title: 'Log In', component: Authenticate };
const NEW_PASSWORD_STEP = { title: 'Set New Password', component: ChangePassword };

export const AUTH_SCHEME = 'USE_AUTHENTICATE_SCHEME';
export const REGISTRATION_SCHEME = 'USE_REGISTRATION_SCHEME';
export const CHANGE_PASSWORD_SCHEME = 'USE_CHANGE_PASSWORD_SCHEME';

const registrationScheme = [
  PHONE_REG_CHECK_STEP,
  PHONE_VERIFICATION_STEP,
  REGISTRATION_STEP,
  CONGRATULATIONS_STEP,
  AUTHENTICATE_STEP,
];

const changePasswordScheme = [PHONE_RECOVERY_CHECK_STEP, PHONE_VERIFICATION_STEP, NEW_PASSWORD_STEP, AUTHENTICATE_STEP];

const authenticateScheme = [AUTHENTICATE_STEP];

const schemes: any = {
  [REGISTRATION_SCHEME]: registrationScheme,
  [CHANGE_PASSWORD_SCHEME]: changePasswordScheme,
  [AUTH_SCHEME]: authenticateScheme,
};

function transformData(state: IAuthSchemeContext, type: string, payload: boolean = true) {
  if (!type) return { ...INITIAL_STATE };
  // Step back
  if (!payload && type === state.schemaName) {
    if (!schemes[type][state.counter - 1]) return { ...state };

    const counter = --state.counter;
    return {
      ...state,
      ...schemes[type][counter],
      counter: counter,
    };
  }
  // End Step back
  // Step forward
  if (type === state.schemaName) {
    if (!schemes[type][state.counter + 1]) return { ...state, ...schemes[type][0], counter: 0 };

    const counter = ++state.counter;

    return {
      ...state,
      ...schemes[type][counter],
      counter: counter,
    };
  }
  // End Step forward
  return {
    ...schemes[type][0],
    counter: 0,
    schemaName: type,
  };
}

export const reducer = (state: IAuthSchemeContext = INITIAL_STATE, { type, payload }: IAction) => {
  if (!!schemes[type]) return transformData(state, type, payload);
  return transformData(state, '');
};

const withSchemaContext = (Component: FC) => {
  function WithSchemaContext(props: any) {
    const [state, dispatch] = useReducer(reducer, INITIAL_STATE);
    const [visible, setVisible] = useState<boolean>(false);

    const handleUpdate = (type: string) => () => {
      if (!!type) {
        dispatch({ type });
        setVisible(true);
        return;
      }
      dispatch({ type: null });
      setVisible(!!type);
    };

    return (
      <AuthSchemeContext.Provider value={{ state, dispatch, visible, handleUpdate }}>
        <Component {...props} />
      </AuthSchemeContext.Provider>
    );
  }
  return WithSchemaContext;
};

export default withSchemaContext;
