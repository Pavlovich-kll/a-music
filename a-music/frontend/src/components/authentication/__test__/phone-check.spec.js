import React from "react";
import Enzyme from 'enzyme';
import {mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import CheckPhoneNumber from "../CheckPhoneNumber";
import { SignUpProvider} from "../../../HOC/withAuthContext";
import {wait} from "@testing-library/react";

Enzyme.configure({adapter: new Adapter()});

describe('Phone-check tests', () => {

    let wrapper;
    let mockHandleClick;

    beforeEach(() => {
        mockHandleClick = jest.fn();
        wrapper = mount(
            <SignUpProvider >
                <CheckPhoneNumber onContinue={mockHandleClick}/>
            </SignUpProvider>);
    });

    it("should render component with initial state", () => {
        expect(wrapper.find('input').length).toBe(1);
        expect(wrapper.find('[data-testid="submit-btn"]').find("button").length).toBe(1);
        expect(wrapper.find('.flag-dropdown').length).toBe(1);
        expect(wrapper.find('.phone-input').instance().value).toEqual('');
    });

    it('should renders property', () => {
        expect(wrapper).toMatchSnapshot()
    });

    it("should change phone-check", async () => {
        const phone = wrapper.find('[name="phone"]').find("input");
        phone.simulate("change", {
            target: {name: "phone", value: "375293192966"}
        });
        wrapper.update();
        await wait(() => {
            expect(phone.instance().value).toEqual('+375 (29) 319 29 66')
        });
    })

});