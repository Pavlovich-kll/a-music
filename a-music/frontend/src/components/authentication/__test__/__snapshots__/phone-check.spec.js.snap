// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Phone-check tests should renders property 1`] = `
<SignUpProvider>
  <CheckPhoneNumber
    onContinue={[MockFunction]}
  >
    <Formik
      initialValues={
        Object {
          "phone": "",
        }
      }
      onSubmit={[Function]}
      validationSchema={
        ObjectSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_defaultDefault": [Function],
          "_deps": Array [],
          "_excludedEdges": Array [],
          "_exclusive": Object {},
          "_mutate": undefined,
          "_nodes": Array [
            "phone",
          ],
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "object",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "fields": Object {
            "phone": StringSchema {
              "_blacklist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "_conditions": Array [],
              "_deps": Array [],
              "_exclusive": Object {
                "matches": false,
                "required": true,
              },
              "_mutate": undefined,
              "_options": Object {
                "abortEarly": true,
                "recursive": true,
              },
              "_type": "string",
              "_typeError": [Function],
              "_whitelist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "tests": Array [
                [Function],
                [Function],
              ],
              "transforms": Array [
                [Function],
              ],
              "type": "string",
            },
          },
          "tests": Array [],
          "transforms": Array [
            [Function],
          ],
          "type": "object",
        }
      }
    >
      <Styled(Form)>
        <Form
          className="sc-AxjAm jarUeW AuthenticateWrappedForm"
        >
          <form
            action="#"
            className="sc-AxjAm jarUeW AuthenticateWrappedForm"
            onReset={[Function]}
            onSubmit={[Function]}
          >
            <Field
              name="phone"
            >
              <t
                alwaysDefaultMask={false}
                areaCodes={null}
                autoFormat={true}
                autocompleteSearch={false}
                buttonClass=""
                buttonStyle={Object {}}
                containerClass=""
                containerStyle={Object {}}
                copyNumbersOnly={true}
                country=""
                countryCodeEditable={true}
                defaultErrorMessage=""
                defaultMask="... ... ... ... .."
                disableCountryCode={false}
                disableDropdown={false}
                disableSearchIcon={false}
                disabled={false}
                dropdownClass=""
                dropdownStyle={Object {}}
                enableAreaCodeStretch={false}
                enableAreaCodes={false}
                enableClickOutside={true}
                enableLongNumbers={false}
                enableSearch={false}
                enableTerritories={false}
                excludeCountries={Array []}
                flagsImagePath="./flags.png"
                inputClass="phone-input"
                inputProps={Object {}}
                inputStyle={Object {}}
                isValid={true}
                jumpCursorToEnd={true}
                keys={
                  Object {
                    "A": 65,
                    "DOWN": 40,
                    "ENTER": 13,
                    "ESC": 27,
                    "LEFT": 37,
                    "PLUS": 43,
                    "RIGHT": 39,
                    "SPACE": 32,
                    "UP": 38,
                    "Z": 90,
                  }
                }
                localization={Object {}}
                masks={null}
                onChange={[Function]}
                onEnterKeyPress={null}
                onlyCountries={Array []}
                placeholder="Enter your phone"
                preferredCountries={Array []}
                prefix="+"
                preserveOrder={Array []}
                priority={null}
                regions=""
                renderStringAsFlag=""
                searchClass=""
                searchNotFound="No entries to show"
                searchPlaceholder="search"
                searchStyle={Object {}}
                showDropdown={false}
                value=""
              >
                <div
                  className=" react-tel-input"
                  onKeyDown={[Function]}
                  style={Object {}}
                >
                  <input
                    className="phone-input form-control"
                    disabled={false}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onClick={[Function]}
                    onCopy={[Function]}
                    onDoubleClick={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    placeholder="Enter your phone"
                    style={Object {}}
                    type="tel"
                    value=""
                  />
                  <div
                    className=" flag-dropdown"
                    role="button"
                    style={Object {}}
                    tabIndex="0"
                  >
                    <div
                      className="selected-flag"
                      onClick={[Function]}
                      title=""
                    >
                      <div
                        className="flag 0"
                      >
                        <div
                          className="arrow"
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </t>
            </Field>
            <WithStyles(ForwardRef(Button))
              color="primary"
              data-testid="submit-btn"
              fullWidth={true}
              type="submit"
              variant="outlined"
            >
              <ForwardRef(Button)
                classes={
                  Object {
                    "colorInherit": "MuiButton-colorInherit",
                    "contained": "MuiButton-contained",
                    "containedPrimary": "MuiButton-containedPrimary",
                    "containedSecondary": "MuiButton-containedSecondary",
                    "containedSizeLarge": "MuiButton-containedSizeLarge",
                    "containedSizeSmall": "MuiButton-containedSizeSmall",
                    "disableElevation": "MuiButton-disableElevation",
                    "disabled": "Mui-disabled",
                    "endIcon": "MuiButton-endIcon",
                    "focusVisible": "Mui-focusVisible",
                    "fullWidth": "MuiButton-fullWidth",
                    "iconSizeLarge": "MuiButton-iconSizeLarge",
                    "iconSizeMedium": "MuiButton-iconSizeMedium",
                    "iconSizeSmall": "MuiButton-iconSizeSmall",
                    "label": "MuiButton-label",
                    "outlined": "MuiButton-outlined",
                    "outlinedPrimary": "MuiButton-outlinedPrimary",
                    "outlinedSecondary": "MuiButton-outlinedSecondary",
                    "outlinedSizeLarge": "MuiButton-outlinedSizeLarge",
                    "outlinedSizeSmall": "MuiButton-outlinedSizeSmall",
                    "root": "MuiButton-root",
                    "sizeLarge": "MuiButton-sizeLarge",
                    "sizeSmall": "MuiButton-sizeSmall",
                    "startIcon": "MuiButton-startIcon",
                    "text": "MuiButton-text",
                    "textPrimary": "MuiButton-textPrimary",
                    "textSecondary": "MuiButton-textSecondary",
                    "textSizeLarge": "MuiButton-textSizeLarge",
                    "textSizeSmall": "MuiButton-textSizeSmall",
                  }
                }
                color="primary"
                data-testid="submit-btn"
                fullWidth={true}
                type="submit"
                variant="outlined"
              >
                <WithStyles(ForwardRef(ButtonBase))
                  className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-fullWidth"
                  component="button"
                  data-testid="submit-btn"
                  disabled={false}
                  focusRipple={true}
                  focusVisibleClassName="Mui-focusVisible"
                  type="submit"
                >
                  <ForwardRef(ButtonBase)
                    className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-fullWidth"
                    classes={
                      Object {
                        "disabled": "Mui-disabled",
                        "focusVisible": "Mui-focusVisible",
                        "root": "MuiButtonBase-root",
                      }
                    }
                    component="button"
                    data-testid="submit-btn"
                    disabled={false}
                    focusRipple={true}
                    focusVisibleClassName="Mui-focusVisible"
                    type="submit"
                  >
                    <button
                      className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-fullWidth"
                      data-testid="submit-btn"
                      disabled={false}
                      onBlur={[Function]}
                      onDragLeave={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      onMouseDown={[Function]}
                      onMouseLeave={[Function]}
                      onMouseUp={[Function]}
                      onTouchEnd={[Function]}
                      onTouchMove={[Function]}
                      onTouchStart={[Function]}
                      tabIndex={0}
                      type="submit"
                    >
                      <span
                        className="MuiButton-label"
                      >
                        Check phone number and continue
                      </span>
                      <WithStyles(memo)
                        center={false}
                      >
                        <ForwardRef(TouchRipple)
                          center={false}
                          classes={
                            Object {
                              "child": "MuiTouchRipple-child",
                              "childLeaving": "MuiTouchRipple-childLeaving",
                              "childPulsate": "MuiTouchRipple-childPulsate",
                              "ripple": "MuiTouchRipple-ripple",
                              "ripplePulsate": "MuiTouchRipple-ripplePulsate",
                              "rippleVisible": "MuiTouchRipple-rippleVisible",
                              "root": "MuiTouchRipple-root",
                            }
                          }
                        >
                          <span
                            className="MuiTouchRipple-root"
                          >
                            <TransitionGroup
                              childFactory={[Function]}
                              component={null}
                              exit={true}
                            />
                          </span>
                        </ForwardRef(TouchRipple)>
                      </WithStyles(memo)>
                    </button>
                  </ForwardRef(ButtonBase)>
                </WithStyles(ForwardRef(ButtonBase))>
              </ForwardRef(Button)>
            </WithStyles(ForwardRef(Button))>
          </form>
        </Form>
      </Styled(Form)>
    </Formik>
  </CheckPhoneNumber>
</SignUpProvider>
`;
